ds/reactive/sorted_set_impl.go

```diff
@@ -0,0 +1,252 @@
+}
+
+// HeaviestElement returns the element with the heaviest weight.
+func (s *sortedSet[ElementType, WeightType]) HeaviestElement() ReadableVariable[ElementType] {
+	return s.heaviestElement
+}
+
+// LightestElement returns the element with the lightest weight.
+func (s *sortedSet[ElementType, WeightType]) LightestElement() ReadableVariable[ElementType] {
+	return s.lightestElement
+}
+
+// addSorted adds the given element to the sortedElements slice.
+func (s *sortedSet[ElementType, WeightType]) addSorted(element ElementType) {
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+
+	if listElement, created := s.elements.GetOrCreate(element, func() *sortedSetElement[ElementType, WeightType] {
+		return newSortedSetElement(element, s)
+	}); created {
+		listElement.unsubscribeFromWeightUpdates = s.weightVariable(element).OnUpdate(func(_ WeightType, newWeight WeightType) {
+			// only lock if this is not the initial update
+			if listElement.unsubscribeFromWeightUpdates != nil {
+				s.mutex.Lock()
+				defer s.mutex.Unlock()
+			}
+
+			listElement.weight = newWeight
+
+			s.updatePosition(listElement)
+		}, true)
+	}
+}
+
+// deleteSorted deletes the given element from the sortedElements slice.
+func (s *sortedSet[ElementType, WeightType]) deleteSorted(element ElementType) {
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+
+	if deletedElement, deleted := s.elements.DeleteAndReturn(element); deleted {
+		// unsubscribe from weight updates
+		deletedElement.unsubscribeFromWeightUpdates()
+
+		// shift all elements to the right of the deleted element one position to the left
+		for i := deletedElement.index; i < len(s.sortedElements)-1; i++ {
+			s.sortedElements[i] = s.sortedElements[i+1]
+			s.sortedElements[i].index--
+		}
+
+		// prevent memory leak and shrink slice
+		s.sortedElements[len(s.sortedElements)-1] = nil
+		s.sortedElements = s.sortedElements[:len(s.sortedElements)-1]
+
+		// update heaviest and lightest element
+		if deletedElement.index == 0 {
+			if len(s.sortedElements) > 0 {
+				s.heaviestElement.Set(s.sortedElements[0].element)
+			} else {
+				s.heaviestElement.Set(*new(ElementType))
+			}
+		}
+		if deletedElement.index == len(s.sortedElements) {
+			if len(s.sortedElements) > 0 {
+				s.lightestElement.Set(s.sortedElements[len(s.sortedElements)-1].element)
+			} else {
+				s.lightestElement.Set(*new(ElementType))
+			}
+		}
+	}
+}
+
+// updatePosition updates the position of the given element in the sortedElements slice.
+func (s *sortedSet[ElementType, WeightType]) updatePosition(element *sortedSetElement[ElementType, WeightType]) (moved bool) {
+	// update heaviest and lightest references after we are done moving the element
+	defer func(oldIndex int) {
+		if moved {
+			if oldIndex == 0 {
+				s.heaviestElement.Set(s.sortedElements[0].element)
+			}
+
+			if oldIndex == len(s.sortedElements)-1 {
+				s.lightestElement.Set(s.sortedElements[len(s.sortedElements)-1].element)
+			}
+		}
+
+		if element.index == 0 {
```

Comment:
No, it is not the same. The oldIndex is the index of the element before it got moved and element.index is the index of the element after it got moved.

So there are 4 cases:

- got moved away from being the heaviest
- got moved away from being the lightest
- is the heaviest
- is the lightest

Since the last two cases should also trigger if the element was not moved (e.g. is the only element in the set), we handle them outside the move check (whenever something becomes the heaviest or lightest it registers itself as such).

The first two cases however only update the heaviest and lightest elements if we got moved away from those spots. In practice it can happen that we set the heaviest / lightest twice but only if there are exactly two elements in the set that swap positions (because then you can "arrive at a heaviest / lightest" spot while simultaneously leaving one).

Capturing these special cases would however require us to consider all possible permutations which would make the code pretty messy. Since the reactive variables already handle the setting of identical values as a noop, I think this is actually simpler than to handle all 16 permutations separately.